drv/transceivers-server/src/main.rs

```diff
@@ -112,6 +158,178 @@ impl ServerImpl {
     }
 }
 
+impl ServerImpl {
+    /// Returns the temperature from a CMIS transceiver.
+    ///
+    /// `port` is a logical port index, i.e. 0-31.
+    fn read_cmis_temperature(
+        &self,
+        port: LogicalPort,
+    ) -> Result<Celsius, FpgaError> {
+        const CMIS_TEMPERATURE_MSB: u8 = 14; // CMIS, Table 8-9
+        self.read_temperature_from_i16(port, CMIS_TEMPERATURE_MSB)
+    }
+
+    /// Returns the temperature from a SFF-8636 transceiver.
+    ///
+    /// `port` is a logical port index, i.e. 0-31.
+    fn read_sff8636_temperature(
+        &self,
+        port: LogicalPort,
+    ) -> Result<Celsius, FpgaError> {
+        const SFF8636_TEMPERATURE_MSB: u8 = 22; // SFF-8636, Table 6-7
+        self.read_temperature_from_i16(port, SFF8636_TEMPERATURE_MSB)
+    }
+
+    /// Trigger a read from the given port's given register, which is assumed to
+    /// be an `i16` containing 1/256 °C.
+    fn read_temperature_from_i16(
+        &self,
+        port: LogicalPort,
+        reg: u8,
+    ) -> Result<Celsius, FpgaError> {
+        self.transceivers.setup_i2c_read(reg, 2, port.as_mask())?;
```

Comment:
At first glance I was surprised this was 2 when `StatusAndTemperature` has 3 bytes worth of data. Is status treated separately even though we pass a `StatusAndTemperature` as a `&mut [u8]`?

Reply:
You had me doubting myself, but I think it's correct:
- `setup_i2c_read` tells us how many bytes to read over I2C (i.e. from the transceiver).  These bytes are written to the I2C read buffer.
- `get_i2c_status_and_read_buffer` reads bytes from the FPGA, so it reads the status byte then two bytes from the I2C read buffer (which is right after it in FPGA register space).

Reply:
I'm sold; not sure how to suggest making this better. Maybe just a comment that the count here is specific to the i16 field and the status is separate?